### PR TITLE
🌱 Generate templates from e2e

### PR DESCRIPTION
**Pull request by mboukhalfa**
The e2e-tests templates were either generated from dev-env or using `envsubst` so the e2e parameters such as clusterName, replicas .. had not any effects.
This PR adds templates with parameters to the e2e tests and remove the dependency on the dev-env template. Also uses `kustomize` to generate OS customized templates.

### CODE REVIEW

**Comment by mboukhalfa** [comment]
/assign @namnx228 
cc @Rozzii

File: Makefile [hunk 1]
```diff
@@ -123,26 +123,35 @@ test-e2e: ## Run e2e tests with capi e2e testing framework
...
 	$(ENVSUBST) < $(E2E_CONF_FILE) > $(E2E_CONF_FILE_ENVSUBST)
 
+## --------------------------------------
+## Templates
+## --------------------------------------
+E2E_TEMPLATES_DIR ?= $(ROOT_DIR)/test/e2e/data/infrastructure-metal3
+.PHONY: cluster-templates
+cluster-templates: $(KUSTOMIZE) ## Generate cluster templates
+	$(KUSTOMIZE) build $(E2E_TEMPLATES_DIR)/cluster-template-ubuntu --load-restrictor LoadRestrictionsNone > $(E2E_TEMPLATES_DIR)/cluster-template-ubuntu.yaml
+	$(KUSTOMIZE) build $(E2E_TEMPLATES_DIR)/cluster-template-centos --load-restrictor LoadRestrictionsNone > $(E2E_TEMPLATES_DIR)/cluster-template-centos.yaml
```
Comment (lentzi90) [inline Makefile:146]: We should not need to change the load restrictor. We can instead change the kustomization (see other comments).
```suggestion
	$(KUSTOMIZE) build $(E2E_TEMPLATES_DIR)/cluster-template-ubuntu > $(E2E_TEMPLATES_DIR)/cluster-template-ubuntu.yaml
	$(KUSTOMIZE) build $(E2E_TEMPLATES_DIR)/cluster-template-centos > $(E2E_TEMPLATES_DIR)/cluster-template-centos.yaml
```

File: Makefile [hunk 2]
```diff
@@ -560,7 +569,6 @@ clean-examples: ## Remove all the temporary files generated in the examples fold
 
 clean-e2e:
 	rm -f $(E2E_CONF_FILE_ENVSUBST)
-	rm -rf $(E2E_ENVSUBST_DIR)
```
Comment (lentzi90) [inline Makefile:563]: Why is this removed?

File: test/e2e/data/infrastructure-metal3/cluster-template-centos/kustomization.yaml [hunk 3]
```diff
@@ -0,0 +1,7 @@
+bases:
+- ../bases/cluster-with-kcp.yaml
+- ../bases/ippool.yaml
+- ../bases/md.yaml
```
Comment (lentzi90) [inline test/e2e/data/infrastructure-metal3/cluster-template-centos/kustomization.yaml:4]: If you put a `kustomize.yaml` file in `bases` with this in it:
```yaml
resources:
- cluster-with-kcp.yaml
- ippool.yaml
- md.yaml
```
we could change this and also get rid of the load restrictor when building.

```suggestion
resources:
- ../base
```

(I changed `bases -> resources` since `bases` is deprecated in kustomize. They basically work the same now.)

File: test/e2e/data/infrastructure-metal3/cluster-template-ubuntu/kustomization.yaml [hunk 4]
```diff
@@ -0,0 +1,7 @@
+bases:
+- ../bases/cluster-with-kcp.yaml
+- ../bases/ippool.yaml
+- ../bases/md.yaml
```
Comment (lentzi90) [inline test/e2e/data/infrastructure-metal3/cluster-template-ubuntu/kustomization.yaml:4]: Same here as in the other `kustomize.yaml`.

File: test/e2e/e2e_test.go [hunk 5]
```diff
@@ -34,7 +34,7 @@ var _ = Describe("Workload cluster creation", func() {
 	BeforeEach(func() {
 		osType := strings.ToLower(os.Getenv("OS"))
 		Expect(osType).ToNot(Equal(""))
-		flavorSuffix = "-m3-dev-env"
+		flavorSuffix = osType
```
Comment (lentzi90) [inline test/e2e/e2e_test.go:37]: It looks like this is no longer a suffix, so maybe change it to just `flavor`?

File: test/e2e/data/infrastructure-metal3/bases/cluster-with-kcp.yaml [hunk 6]
```diff
@@ -0,0 +1,117 @@
+apiVersion: cluster.x-k8s.io/${CAPI_VERSION}
+kind: Cluster
+metadata:
+  name: ${CLUSTER_NAME}
+  namespace: ${NAMESPACE}
```
Comment (lentzi90) [inline test/e2e/data/infrastructure-metal3/bases/cluster-with-kcp.yaml:5]: I think we can remove the namespace here and put it in the `kustomize.yaml` file instead.

**Review by lentzi90 — COMMENTED** [review]
I like the use of kustomize to handle the duplication between centos and ubuntu templates!
The use of envsubst on the other hand I'm not so fond of. I was thinking that we could maybe use go templates instead? That way we could use the same variables we use in the go code to render the templates. It is not a strong preference but I would like to hear what others think.

File: test/e2e/data/infrastructure-metal3/cluster-template-centos/kustomization.yaml (hunk 3 above)
Comment (mboukhalfa) [reply test/e2e/data/infrastructure-metal3/cluster-template-centos/kustomization.yaml:4]: At some point I think we might not need ippool.yaml that was the case in the upgrade (test workload template)

File: Makefile (hunk 1 above)
Comment (mboukhalfa) [reply Makefile]: What do you mean by changing the kustomization

File: Makefile [hunk 7]
```diff
@@ -123,26 +123,35 @@ test-e2e: ## Run e2e tests with capi e2e testing framework
...
 .PHONY: e2e-substitutions
-e2e-substitutions: $(ENVSUBST) $(subst $(E2E_TEMPLATES_DIR),$(E2E_ENVSUBST_DIR),$(wildcard $(E2E_TEMPLATES_DIR)/*.yaml))
+e2e-substitutions: $(ENVSUBST)
+	mkdir -p $(E2E_ENVSUBST_DIR)
 	$(ENVSUBST) < $(E2E_CONF_FILE) > $(E2E_CONF_FILE_ENVSUBST)
 
+## --------------------------------------
+## Templates
+## --------------------------------------
+E2E_TEMPLATES_DIR ?= $(ROOT_DIR)/test/e2e/data/infrastructure-metal3
```
[the remark this answers is not in the document]
Comment (mboukhalfa) [reply Makefile:146]: Done also `E2E_ENVSUBST_DIR` renamed to `E2E_OUT_DIR`

File: hack/e2e/environment.sh [hunk 8]
```diff
@@ -58,3 +58,7 @@ source "${M3_DEV_ENV_PATH}/scripts/feature_tests/node_reuse/node_reuse_vars.sh"
 # Pin Calico version
 export CALICO_PATCH_RELEASE="v3.21.0"
 export CALICO_MINOR_RELEASE="v3.21"
+
+# Exported to the cluster templates
+SSH_PUB_KEY_CONTENT=$(cat "${HOME}/.ssh/id_rsa.pub")
+export SSH_PUB_KEY_CONTENT
```
[the remark this answers is not in the document]
Comment (mboukhalfa) [reply hack/e2e/environment.sh]: Moved to Makefile

File: Makefile (hunk 2 above)
Comment (mboukhalfa) [reply Makefile:563]: Thanks, the line `rm -f $(E2E_CONF_FILE_ENVSUBST)` removed instead of `rm -rf $(E2E_ENVSUBST_DIR)` since it is a child file inside the Dir

File: test/e2e/e2e_test.go (hunk 5 above)
Comment (mboukhalfa) [reply test/e2e/e2e_test.go]: changed

File: test/e2e/data/infrastructure-metal3/bases/cluster-with-kcp.yaml (hunk 6 above)
Comment (mboukhalfa) [reply test/e2e/data/infrastructure-metal3/bases/cluster-with-kcp.yaml:5]: As a result the namespace will be still a parameter do not see the reason for that

File: test/e2e/data/infrastructure-metal3/cluster-template-ubuntu/kustomization.yaml (hunk 4 above)
Comment (mboukhalfa) [reply test/e2e/data/infrastructure-metal3/cluster-template-ubuntu/kustomization.yaml:4]: see the comment

File: test/e2e/config/e2e_conf.yaml [hunk 9]
```diff
@@ -62,16 +62,12 @@ providers:
     type: "url"
     contract: v1beta1
     files:
-    - sourcePath: "${PWD}/metadata.yaml"
+    - sourcePath: "../../../metadata.yaml"
```
Comment (namnx228) [inline test/e2e/config/e2e_conf.yaml:65]: This metadata.yaml file that e2e uses is different from the one in the root directory. Please revert this change

**Review by namnx228 — CHANGES_REQUESTED** [review]
This is a really good work. Just one thing to consider. See my comment

File: test/e2e/config/e2e_conf.yaml (hunk 9 above)
Comment (mboukhalfa) [reply test/e2e/config/e2e_conf.yaml:65]: if you mean reverting `${PWD}/metadata.yaml` it will end up using the same file

File: test/e2e/data/infrastructure-metal3/bases/cluster-with-kcp.yaml (hunk 6 above)
Comment (lentzi90) [reply test/e2e/data/infrastructure-metal3/bases/cluster-with-kcp.yaml:5]: It is just simpler. One place instead of many. Now we need to have it for the Cluster, Metal3Cluster, KubeadmControlPlane and Metal3MachineTemplates. If we put it in `kustomize.yaml` it is only in one place and applies to all of them I think.

File: test/e2e/data/infrastructure-metal3/cluster-template-centos/kustomization.yaml (hunk 3 above)
Comment (lentzi90) [reply test/e2e/data/infrastructure-metal3/cluster-template-centos/kustomization.yaml:4]: In that case we should have a separate base for the IP pool. Something like this:
```
├── bases
│   ├── cluster
│   │   ├── cluster-with-kcp.yaml
│   │   ├── kustomization.yaml
│   │   └── md.yaml
│   └── ippool
│       ├── ippool.yaml
│       └── kustomization.yaml
└── overlays
    ├── centos
    │   └── kustomization.yaml
    └── ubuntu
        └── kustomization.yaml
```

File: Makefile (hunk 1 above)
Comment (lentzi90) [reply Makefile]: It is the [other comment](https://github.com/metal3-io/cluster-api-provider-metal3/pull/537#discussion_r814644367)

File: test/e2e/data/infrastructure-metal3/bases/cluster/kustomization.yaml [hunk 10]
```diff
@@ -0,0 +1,3 @@
+bases:
```
Comment (lentzi90) [inline test/e2e/data/infrastructure-metal3/bases/cluster/kustomization.yaml:1]: Nit.
```suggestion
resources:
```

File: test/e2e/data/infrastructure-metal3/bases/ippool/kustomization.yaml [hunk 11]
```diff
@@ -0,0 +1,2 @@
+bases:
```
Comment (lentzi90) [inline test/e2e/data/infrastructure-metal3/bases/ippool/kustomization.yaml:1]:
```suggestion
resources:
```

File: test/e2e/data/infrastructure-metal3/cluster-template-centos/kustomization.yaml [hunk 12]
```diff
@@ -0,0 +1,6 @@
+bases:
```
Comment (lentzi90) [inline test/e2e/data/infrastructure-metal3/cluster-template-centos/kustomization.yaml:1]:
```suggestion
resources:
```

File: test/e2e/data/infrastructure-metal3/cluster-template-ubuntu/kustomization.yaml [hunk 13]
```diff
@@ -0,0 +1,6 @@
+bases:
```
Comment (lentzi90) [inline test/e2e/data/infrastructure-metal3/cluster-template-ubuntu/kustomization.yaml:1]:
```suggestion
resources:
```

**Review by lentzi90 — COMMENTED** [review]
Nice!
I added suggestions to change `bases` to `resources`, since it [has been deprecated](https://github.com/kubernetes-sigs/kustomize/blob/661743c7e5bd8c3d9d6866b6bc0a6f0e0b0512eb/site/content/en/blog/releases/v2.1.0.md#field-changes--deprecations). We are still on an old version of kustomize where it is supported but it is good to prepare already for the future.

Other than that I think it is good to go in! :slightly_smiling_face:

File: test/e2e/config/e2e_conf.yaml [hunk 14]
```diff
@@ -87,7 +83,8 @@ variables:
   CAPIRELEASE: "${CAPIRELEASE}"
   CAPM3RELEASE: "${CAPM3RELEASE}"
   CONFIG_FILE_PATH: "${HOME}/.cluster-api/overrides/clusterctl.yaml"
-
+  # Default vars for the template, those values could be overridden by the env-vars.
+  SERVICE_CIDR: "10.96.0.0/12"
```
[the remark this answers is not in the document]
Comment (mboukhalfa) [reply test/e2e/config/e2e_conf.yaml:87]: I do not think it is exported in the env because e2e was complaining about this var while generating the template but I can add possibility to read from env else use the default value

[the remark this answers is not in the document]
Comment (mboukhalfa) [reply test/e2e/config/e2e_conf.yaml:87]: Well I guess e2e will prioritize the value in the env else it use what is declared in conf file so no need to read from env in the conf file

**Review by lentzi90 — APPROVED** [review]

**Review by furkatgofurov7 — COMMENTED** [review]
Good job, thanks! 
The only thing I would suggest is to rename file names to be specific:

- md-KubeadmConfigTemplate.yaml ==> workers-kubeadm-config.yaml 
- kcp-KubeadmControlPlane.yaml ==> controlplane-kubeadm-config.yaml

File: test/e2e/data/infrastructure-metal3/bases/ippool/kustomization.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (mboukhalfa) [reply test/e2e/data/infrastructure-metal3/bases/ippool/kustomization.yaml]: done

**Review by furkatgofurov7 — APPROVED** [review]
/retitle 🌱 Generate templates from e2e
/approve
